Mathlib/Analysis/Calculus/Gradient.lean

```diff
@@ -0,0 +1,472 @@
+  let f' := ∇* gradf
+  let g' := ∇* gradg
+  have h₁: HasFDerivAt (g ∘ f) (g'.comp f') x:= by
+    apply HasFDerivAt.comp
+    rw [HasGradoneAt] at hg; apply hg
+    apply hf
+  have h₂ : (∇* (gradg • gradf)) = g'.comp f':= by
+    apply ContinuousLinearMap.ext_iff.mpr; intro y
+    have l₁: (g'.comp f') y = g' (f' y) := rfl
+    have l₂: f' y = ⟪gradf, y⟫ := rfl
+    have l₃: g' ⟪gradf, y⟫ = gradg * ⟪gradf, y⟫ := rfl
+    have l₄: (∇* (gradg • gradf)) y = ⟪gradg • gradf, y⟫ := rfl
+    rw [l₁, l₂, l₃, l₄]
+    exact real_inner_smul_left gradf y gradg
+  rw [HasGradnAt, h₂]
+  apply h₁
+
+theorem HasGradnAt.mul (hf : HasGradnAt f gradf x) (hg : HasGradnAt g gradg x) :
+    HasGradnAt (fun y => f y * g y) (f x • gradg + g x • gradf) x := by
+  let f':= ∇* gradf
+  let g':= ∇* gradg
+  have h₁: HasFDerivAt (fun x => f x * g x) (f x • g' + g x • f') x:= by
+    apply HasFDerivAt.mul hf hg
+  have h₂: f x • g' + g x • f' =  (∇* (f x • gradg + g x • gradf)):= by
+    apply ContinuousLinearMap.ext_iff.mpr; intro y
+    have l₁: (f x • g' + g x • f') y = (f x • g') y + (g x • f') y:= rfl
+    have l₂: (f x • g') y + (g x • f') y = f x * (g' y) + g x * (f' y) := rfl
+    have l3: f x * (g' y) + g x * (f' y) = f x * ⟪gradg, y⟫ + g x * ⟪gradf, y⟫ := rfl
+    have l4: f x * ⟪gradg, y⟫ + g x * ⟪gradf, y⟫ = ⟪f x • gradg, y⟫ + ⟪g x • gradf, y⟫ := by
+      rw [real_inner_smul_left gradg y (f x), real_inner_smul_left gradf y (g x)]
+    have l5: ⟪f x • gradg, y⟫ + ⟪g x • gradf, y⟫ = ⟪f x • gradg + g x • gradf, y⟫:= by
+      rw [inner_add_left]
+    rw [l₁, l₂, l3, l4, l5]
+    rfl
+  rw [HasGradnAt, ← h₂]
+  apply h₁
+
+theorem toDual.unique {f₁ g₁: E} (h: (∇* f₁) = (∇* g₁)): f₁ = g₁ := by
+  have h₁: ∀ x : E , (∇* f₁) x = (∇* g₁) x :=
+   fun x => congrFun (congrArg FunLike.coe h) x
+  specialize h₁ (f₁ - g₁)
+  have : (∇* f₁) (f₁ - g₁) = ⟪f₁, f₁ - g₁⟫ := rfl
+  rw [this] at h₁
+  have : (∇* g₁) (f₁ - g₁) = ⟪g₁, f₁ - g₁⟫ := rfl
+  rw [this] at h₁
+  have : ⟪f₁, f₁ - g₁⟫ - ⟪g₁, f₁ - g₁⟫ = 0 := by
+    rw [h₁, sub_self]
+  rw [← inner_sub_left, inner_self_eq_zero, sub_eq_zero] at this
+  exact this
+
+theorem HasGradnAt.unique (hf : HasGradnAt f gradf x) (hg : HasGradnAt f gradg x):
+  gradf = gradg := by
+  rw [HasGradnAt] at hf
+  rw [HasGradnAt] at hg
+  have : (∇* gradf) = (∇* gradg):= by
+    apply HasFDerivAt.unique hf hg
+  apply toDual.unique this
+
+theorem HasGradnAt.congr (hf : HasGradnAt f gradf x) (hg : HasGradnAt g gradg x) (equal: f = g):
+  gradf = gradg := by
+  rw [equal] at hf
+  apply HasGradnAt.unique hf hg
+
+theorem HasGradnAt.constant (c : ℝ ) : HasGradnAt (fun (_ : E) => c) 0 x := by
+  rw [HasGradnAt_iff_Convergence]
+  intro ε epos
+  use (1 : ℝ); constructor;
+  · exact Real.zero_lt_one
+  · intro x' _
+    rw [sub_self, inner_zero_left, sub_self, norm_zero]
+    apply mul_nonneg (LT.lt.le epos) (norm_nonneg (x - x'))
+
+/-
+  The following theorems concerns the relationship of
+  `ε - δ` convergence, gradient and continuity.
```

Comment:
Make these exact copies of similar results for `HasFDerivAt` (and use filter notation), and prove them directly from those results.